server/common/default_config.py

```diff
@@ -1,129 +1,148 @@
-  #         set2 : /path/to/set2_datasets/
-  # In this case, datasets can be accessed from <server>/set1/<datasetname> or
-  # <server>/set2/<datasetname>.
-
-  dataroot: null
-
-  # The index page when in multi-dataset mode:
-  #   false or null:  this returns a 404 code
-  #   true:  loads a test index page, which links to the datasets that are available in the dataroot
-  #   string/URL:  redirect to this URL:  flask.redirect(config.multi_dataset__index)
-  index: false
-
-  # A list of allowed matrix types.  If an empty list, then all matrix types are allowed
-  allowed_matrix_types: []
-
-  matrix_cache:
-    # The maximum number of datasets that may be opened at one time.  The least recently used dataset
-    # is evicted from the cache first.
-    max_datasets: 5
-
-    # A matrix is automatically removed from the cache after timelimit_s number of seconds.
-    # If timelimit_s is set to None, then there is no time limit.
-    timelimit_s: 30
-
-single_dataset:
-  datapath: null
-  obs_names: null
-  var_names: null
-  about: null
-  title: null
-
-user_annotations:
-  enable: true
-  type: local_file_csv
-  local_file_csv:
-    directory: null
-    file: null
-  ontology:
-    enable: false
-    obo_location: null
-
-embeddings:
-  names : []
-  enable_reembedding: false
-
-diffexp:
-  enable: true
-  lfc_cutoff: 0.01
-  top_n: 10
-  alg_cxg:
-    # The number of threads to use is computed from: min(max_workers, cpu_multipler * cpu_count).
-    # Where cpu_count is determined at runtime.
-    max_workers: 64
-    cpu_multiplier: 4
-
-    # The target number of matrix elements that are evaluated
-    # together in one thread.
-    target_workunit: 16_000_000
-
-data_locator:
-  s3:
-    # s3 region name.
-    #   if true, then the s3 location is automatically determined from the datapath or dataroot.
-    #   if false/null, then do not set.
-    #   if a string, then use that value (e.g. us-east-1).
-    region_name: true
-
-adaptor:
-  cxg_adaptor:
-    # The key/values under tiledb_ctx will be used to initialize the tiledb Context.
-    # If 'vfs.s3.region' is not set, then it will automatically use the setting from
-    # data_locator / s3 / region_name.
-    tiledb_ctx:
-      sm.tile_cache_size:  8589934592
-      sm.num_reader_threads:  32
-
-  anndata_adaptor:
+  app:
+    verbose: false
+    debug: false
+    host: "127.0.0.1"
+    port : null
+    open_browser: false
+    force_https: false
+    flask_secret_key: null
+    generate_cache_control_headers: false
+    server_timing_headers: false
+    csp_directives: null
+
+  multi_dataset:
+    # If dataroot is set, then cellxgene may serve multiple datasets.  This parameter is not
+    # compatible with single_dataset/datapath.
+    # dataroot may be a string, representing the path to a directory or S3 prefix.  In this
+    # case the datasets in that location are accessed from <server>/d/<datasetname>.
+    # example:
+    #     dataroot:  /path/to/datasets/
```

Comment:
If I understand correctly, this is a shorthand for:

```
dataroot:
    d: /path/to/datasets/
```

If that is true, would it simplify things to EOL the shortcut and only support the dictionary config syntax (ie, the long form)?  It would have the benefit of removing the magic "/d/" from the code base.

Reply:
Yes, I could do that.   One small issue is that "--dataroot" is a command line option, which is pretty handy.  I could either 1) remove the --dataroot command line option, which means that a to setup a multi dataset local server you need to have a config file or 2).  keep it, and replace it with the dictionary, { "d" : dataroot },  when evaluating the config options.     In either case it gets rid of the special handling in the app.py code.  In the second case it moves the special handling to app_config.py.  The special handling in app_config.py would be pretty trivial.

Reply:
I think the CLI params could be left dirt simple, and only cover a subset of what you can do in the config.  So in this case (`--dataroot`) it would just be a shortcut...ie, your second option.